plugins/scaffolder-backend/src/scaffolder/actions/builtin/github/githubEnvironment.ts

```diff
@@ -0,0 +1,206 @@
+              custom_branch_policies: {
+                title: 'Custom Branch Policies',
+                description: `Whether only branches that match the specified name patterns can deploy to this environment. If custom_branch_policies is true, protected_branches must be false; if custom_branch_policies is false, protected_branches must be true.`,
+                type: 'boolean',
+              },
+            },
+          },
+          customBranchPolicyNames: {
+            title: 'Custom Branch Policy Name',
+            description: `The name pattern that branches must match in order to deploy to the environment.
+
+            Wildcard characters will not match /. For example, to match branches that begin with release/ and contain an additional single slash, use release/*/*. For more information about pattern matching syntax, see the Ruby File.fnmatch documentation.`,
+            type: 'array',
+            items: {
+              type: 'string',
+            },
+          },
+          environmentVariables: {
+            title: 'Environment Variables',
+            description: `Environment variables attached to the deployment environment`,
+            type: 'object',
+          },
+          secrets: {
+            title: 'Deployment Secrets',
+            description: `Secrets attached to the deployment environment`,
+            type: 'object',
+          },
+          token: {
+            title: 'Authentication Token',
+            type: 'string',
+            description: 'The token to use for authorization to GitHub',
+          },
+        },
+      },
+    },
+    async handler(ctx) {
+      const {
+        repoUrl,
+        name,
+        deploymentBranchPolicy,
+        customBranchPolicyNames,
+        environmentVariables,
+        secrets,
+        token: providedToken,
+      } = ctx.input;
+
+      const octokitOptions = await getOctokitOptions({
+        integrations,
+        token: providedToken,
+        repoUrl: repoUrl,
+      });
+
+      const { owner, repo } = parseRepoUrl(repoUrl, integrations);
+
+      if (!owner) {
+        throw new InputError(`No owner provided for repo ${repoUrl}`);
+      }
+
+      const client = new Octokit(octokitOptions);
+      const repository = await client.rest.repos.get({
+        owner: owner,
+        repo: repo,
+      });
+
+      await client.rest.repos.createOrUpdateEnvironment({
+        owner: owner,
+        repo: repo,
+        environment_name: name,
+        deployment_branch_policy: deploymentBranchPolicy ?? null,
+      });
+
+      if (customBranchPolicyNames) {
+        for (const item of customBranchPolicyNames) {
+          await client.rest.repos.createDeploymentBranchPolicy({
+            owner: owner,
+            repo: repo,
+            environment_name: name,
+            name: item,
+          });
+        }
+      }
+
+      for (const key in environmentVariables) {
+        if ({}.hasOwnProperty.call(environmentVariables, key)) {
```

Comment:
Same as in #17722 - would prefer `for (const [key, value] of Object.entries(...))` if possible, same below